### PR TITLE
aws-lib version upgrade + pagination support

### PR DESCRIPTION
**Pull request by oren-zohar**
### In this PR:
- Updating `aws-sdk-go-v2` version to 1.16.7
- Pagination support for `DescribeRepositories` and `DescribeLoadBalancer`
- General code refactoring
____
- fixes: https://github.com/elastic/cloudbeat/issues/283
- fixes: https://github.com/elastic/security-team/issues/3175

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -45,7 +52,20 @@ require (
 	k8s.io/api v0.24.2
 )
 
-require github.com/dschmidt/go-layerfs v0.1.0 // indirect
+require (
```
Comment (ofiriro3) [inline go.mod:55]: Wondering about this `require` section.

I see that we didn't import the `internal\service` this way before.

We used to have only a single import of `github.com/aws/aws-sdk-go-v2 v0.24.0`

WDYT?

File: resources/providers/awslib/eks_provider.go [hunk 2]
```diff
@@ -23,32 +23,35 @@ import (
...
-	response, err := req.Send(ctx)
+
+	response, err := provider.client.DescribeCluster(ctx, input)
 	if err != nil {
-		return nil, fmt.Errorf("failed to describe cluster %s from eks , error - %w", clusterName, err)
+		return EksCluster{}, fmt.Errorf("failed to describe cluster %s from eks, error - %w", clusterName, err)
 	}
 
-	return response, err
+	return EksCluster(*response.Cluster), err
```
Comment (ofiriro3) [inline resources/providers/awslib/eks_provider.go:56]: I suspect it is a breaking change, and we need to address it in the CSP rules as well/reverting it

For example, https://github.com/elastic/csp-security-policies/blob/main/bundle/compliance/cis_eks/rules/cis_2_1_1/rule.rego

We used to access the cluster field by using `resource.Cluster`

File: resources/providers/awslib/ecr_provider.go [hunk 3]
```diff
@@ -20,45 +20,51 @@ package awslib
...
 	return &ECRProvider{
 		client: svc,
 	}
 }
 
 // DescribeAllECRRepositories This method will return a maximum of 100 repository
-/// If we will ever wish to change it, DescribeRepositories returns results in paginated manner
-func (provider *ECRProvider) DescribeAllECRRepositories(ctx context.Context) ([]ecr.Repository, error) {
-	/// When repoNames is nil, it will describe all the existing repositories
+// If we will ever wish to change it, DescribeRepositories returns results in paginated manner
```
Comment (ofiriro3) [inline resources/providers/awslib/ecr_provider.go:46]: No need for this comment anymore

File: resources/providers/awslib/ecr_provider.go [hunk 4]
```diff
@@ -20,45 +20,51 @@ package awslib
...
+	for paginator.HasMorePages() {
+		page, err := paginator.NextPage(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("error DescribeRepositories with Paginator: %w", err)
+		}
+		repos = append(repos, page.Repositories...)
 	}
 
-	return response.Repositories, err
+	return repos, nil
```
Comment (ofiriro3) [inline resources/providers/awslib/ecr_provider.go:69]: It's good for a first iteration solution, but in the long term, I think that we might want to consider doing it on the infrastructure level.

**Review by ofiriro3 — CHANGES_REQUESTED** [review]
Huge work + a lot of refactoring 💯 

I found a small breaking change that is critical and we have to address it, the rest is some small comments.

But I'd like to go over it from the beginning just to make sure I hadn't miss anything

File: resources/providers/awslib/eks_provider.go (hunk 2 above)
Comment (oren-zohar) [reply resources/providers/awslib/eks_provider.go:56]: yes we discussed it and you said you're ok with it, I can revert though

File: go.mod [hunk 5]
```diff
@@ -239,7 +259,6 @@ replace (
 	github.com/Microsoft/go-winio => github.com/bi-zone/go-winio v0.4.15
 	github.com/Shopify/sarama => github.com/elastic/sarama v1.19.1-0.20210823122811-11c3ef800752
 	github.com/apoydence/eachers => github.com/poy/eachers v0.0.0-20181020210610-23942921fe77 //indirect, see https://github.com/elastic/beats/pull/29780 for details.
-	github.com/cucumber/godog => github.com/cucumber/godog v0.8.1
```
Comment (ofiriro3) [inline go.mod:242]: Was it removed on purpose?

File: resources/fetchers/ecr_fetcher.go [hunk 6]
```diff
@@ -56,7 +56,7 @@ type ECRFetcherConfig struct {
 	KubeConfig                    string `config:"Kubeconfig"`
 }
 
-type EcrRepository ecr.Repository
+type EcrRepository types.Repository
```
Comment (ofiriro3) [inline resources/fetchers/ecr_fetcher.go:59]: To completely finish the amazing segregation that you made, can we replace the `types. Repository` with a type of our own? 

Same as you did in the IAM and the EKS fetchers

That way, only the AWS provider will be aware of the AWS interface

File: resources/fetchers/ecr_fetcher_test.go [hunk 7]
```diff
@@ -196,8 +195,8 @@ func (s *ECRFetcherTestSuite) TestCreateFetcher() {
 					Name:  "cloudbeat1",
 				},
 			},
-			[]ecr.Repository{},
-			[]ecr.Repository{},
+			awslib.EcrRepositories{},
```
Comment (ofiriro3) [inline resources/fetchers/ecr_fetcher_test.go:198]: Perfect ! the test only interacts with the awslib interface!

File: resources/fetchers/eks_fetcher.go [hunk 8]
```diff
@@ -24,50 +24,49 @@ import (
...
-type EKSFetcherConfig struct {
+type EksFetcherConfig struct {
 	fetching.AwsBaseFetcherConfig `config:",inline"`
 	ClusterName                   string `config:"clusterName"`
 }
 
-type EKSResource struct {
-	*eks.DescribeClusterResponse
+type EksResource struct {
+	awslib.EksClusterOutput
```
Comment (ofiriro3) [inline resources/fetchers/eks_fetcher.go:43]: 🚀

File: resources/fetchers/elb_fetcher.go [hunk 9]
```diff
@@ -51,7 +50,7 @@ type ELBFetcherConfig struct {
 	KubeConfig                    string `config:"Kubeconfig"`
 }
 
-type LoadBalancersDescription elasticloadbalancing.LoadBalancerDescription
+type LoadBalancersDescription types.LoadBalancerDescription
```
Comment (ofiriro3) [inline resources/fetchers/elb_fetcher.go:53]: Can we replace `types.LoadBalancerDescription` with `awslib.***`? 

Same way as you did for EKS fetcher and the IAM fetcher?

File: resources/providers/awslib/eks_provider.go [hunk 10]
```diff
@@ -20,35 +20,39 @@ package awslib
...
 	"context"
 	"fmt"
+	"github.com/aws/aws-sdk-go-v2/service/eks/types"
 
 	"github.com/aws/aws-sdk-go-v2/aws"
 	"github.com/aws/aws-sdk-go-v2/service/eks"
 )
 
+type EksClusterOutput eks.DescribeClusterOutput
+type EksCluster types.Cluster
```
Comment (ofiriro3) [inline resources/providers/awslib/eks_provider.go:30]: I think that we don't use the field anymore, are we?

**Review by ofiriro3 — CHANGES_REQUESTED** [review]

File: go.mod (hunk 5 above)
Comment (oren-zohar) [reply go.mod:242]: yes, we don't use it

File: resources/providers/awslib/eks_provider.go (hunk 10 above)
Comment (oren-zohar) [reply resources/providers/awslib/eks_provider.go:30]: we are using it in the tests

File: resources/providers/awslib/elb_provider.go [hunk 11]
```diff
@@ -26,6 +26,7 @@ import (
 )
 
 type ElbLoadBalancerDescriptions []types.LoadBalancerDescription
```
Comment (ofiriro3) [inline resources/providers/awslib/elb_provider.go:28]: nit: isn't the `ElbLoadBalancerDescriptions` an array of `ElbLoadBalancersDescription`?

Do we need them both?

File: resources/providers/awslib/ecr_provider.go [hunk 12]
```diff
@@ -25,6 +25,7 @@ import (
 	"github.com/aws/aws-sdk-go-v2/service/ecr/types"
 )
 
+type EcrRepository types.Repository
```
Comment (ofiriro3) [inline resources/providers/awslib/ecr_provider.go:28]: nit: isn't the `EcrRepositories` an array of `EcrRepository`?

Do we need them both?

**Review by ofiriro3 — APPROVED** [review]
No further review is necessary! Amazing job

But please take a look at my last two comments 🥇

**Pull request merged**